### PR TITLE
New and improved IK system for Skeleton2D - Squished

### PR DESCRIPTION
**Pull request by TwistedTwigleg**
This PR and commit adds a new IK system for 2D with the Skeleton2D node that adds several new IK solvers, a way to control bones in a Skeleton2D node similar to that in Skeleton3D. It also adds additional changes and functionality.

This work was sponsored by GSoC 2020 and TwistedTwigleg.
* Original, non-squished GitHub PR: https://github.com/godotengine/godot/pull/40347
* GSoC 2020 link: https://summerofcode.withgoogle.com/archive/2020/projects/6588671165202432/

**Once merged, the code in this PR will be up to the community to maintain.** I will help advise and contribute small changes infrequently as best I can, but I cannot take the responsibility of maintaining and overseeing the code.

**Huge thanks to everyone that has helped with this PR and getting the PR to this point!** :tada:

### CODE REVIEW

**Comment by pouleyKetchoupp** [comment]
The error in different builds is (this one is the windows version):
```
.\scene/2d/physical_bone_2d.h(82): error C3668: 'PhysicalBone2D::get_configuration_warning': method with override specifier 'override' did not override any base class methods
```

But I'm not sure what changed. Maybe a rebase went wrong?

**Comment by TwistedTwigleg** [comment]
> The error in different builds is (this one is the windows version):
> 
> ```
> .\scene/2d/physical_bone_2d.h(82): error C3668: 'PhysicalBone2D::get_configuration_warning': method with override specifier 'override' did not override any base class methods
> ```
> 
> But I'm not sure what changed. Maybe a rebase went wrong?

Possibly something went wrong with the rebase, but the non-squished version also had errors with the CI and it was likely the same thing. The only thing that changed though from the last successful CI build was changing the `looking_at` function, which shouldn't have influenced this at all.

Looking at how other nodes override the function in their header, it uses the same syntax and formatting as they do, so I'm not sure why it isn't working here when it works there.

**Comment by pouleyKetchoupp** [comment]
It's actually due to #43180, there's no conflict but you would have the same warnings if you rebased locally.

This method now returns an array of strings instead of one string with line breaks.

**Comment by TwistedTwigleg** [comment]
The CI issue should be fixed. I rebased the PR with the latest version of `master` and fixed the issue with the `get_configuration_warning` function.

However, there is an issue with the physical_bone_2d node that now that causes a segmentation fault. According to the stack trace, it is caused by the `is_node_in_tree` function that is called when calling `get_bone` on the Skeleton2D node. I tried to do some quick fixes but they did not work. I will look into the issue further once I am able.

**Comment by TwistedTwigleg** [comment]
The segmentation fault issue has been fixed! Turns out it was an issue where the PhysicsBone2D was positioning itself at the Bone2D position in the Godot editor at start up. I removed the notification and it works fine. Removing it also shouldn't change anything functionally, the node can just now be moved around in the editor but it will snap itself into place when running.

Now everything should be fixed and good to go

___________

There is one minor thing, but I'm not sure it is related to this PR, is that the PhysicsBone2D spits out a bunch of the following error when run:

~~~
ERROR: Condition "det == 0" is true.
   at: affine_invert (core/math/transform_2d.cpp:49)
~~~

Which seems to be caused by line `70` in the `_position_at_bone2d` function, specifically in the `set_global_transform` function call. My only guess is that the parent item has a determinate of 0, but I'm not sure why it is spitting an error out when it didn't before.
I am not really sure what I can do in  PhysicsBone2D to fix the issue without altering the `set_global_transform` function or the `affine_inverse` function in `Transform2D`.

**Comment by pouleyKetchoupp** [comment]
About the `affine_invert` error:

I'm able to reproduce it when running this scene from your test project:
https://github.com/TwistedTwigleg/Godot_GSOC_2020_2D_Project/tree/master/Scenes/Test_09

And I'm getting this callstack in debug:
```
 godot.windows.tools.64.exe!Transform2D::affine_invert() Line 49	C++
 godot.windows.tools.64.exe!Transform2D::affine_inverse() Line 63	C++
 godot.windows.tools.64.exe!PinJoint2DSW::setup(float p_step) Line 142	C++
 godot.windows.tools.64.exe!Step2DSW::_setup_island(LocalVector<Constraint2DSW *,unsigned int,0> & p_constraint_island, float p_delta) Line 70	C++
 godot.windows.tools.64.exe!Step2DSW::step(Space2DSW * p_space, float p_delta, int p_iterations) Line 218	C++
 godot.windows.tools.64.exe!PhysicsServer2DSW::step(float p_step) Line 1263	C++
 godot.windows.tools.64.exe!PhysicsServer2DWrapMT::step(float p_step) Line 76	C++
 godot.windows.tools.64.exe!Main::iteration() Line 2488	C++
 godot.windows.tools.64.exe!OS_Windows::run() Line 622	C++
```

So if it's related to the case you spotted, it's due to setting up a joint between a kinematic body and a static body (infinite inertia and mass are leading to a zeroed matrix). I'm not sure why this would have changed, but maybe the error wasn't spotted on version 3.2. Also it looks like the error wouldn't be raised in release_debug.

In any case, I think this can be solved outside of this PR. It probably makes sense for the physics server to handle this case by just ignoring the joint, so I'll open a PR for that.

**Review by pouleyKetchoupp — APPROVED** [review]

**Comment by akien-mga** [comment]
Welp, looks like GitHub closes the whole PR when merging another one that fixes a *comment*.

**Comment by akien-mga** [comment]
It was good in the end that this got closed and reopened as it restarted CI, and there's a new issue to handle: https://github.com/godotengine/godot/pull/47872/checks?check_run_id=2359849305

We merged https://github.com/godotengine/godot/pull/47414 yesterday so we now have @qarmin's `RegressionTestProject` running on CI and it seems to get a crash with this PR:

```
#################### SkeletonModification2DPhysicalBones ####################
SkeletonModification2DPhysicalBones.set_physical_bone_chain_length
Parameters [100]
SkeletonModification2DPhysicalBones.get_physical_bone_chain_length
Parameters []
SkeletonModification2DPhysicalBones.set_physical_bone_node
Parameters [100, ]
ERROR: Joint index out of range!
   at: set_physical_bone_node (scene/resources/skeleton_modification_2d_physicalbones.cpp:259)
SkeletonModification2DPhysicalBones.get_physical_bone_node
Parameters [100]
ERROR: Joint index out of range!
   at: get_physical_bone_node (scene/resources/skeleton_modification_2d_physicalbones.cpp:265)
SkeletonModification2DPhysicalBones.fetch_physical_bones
Parameters []
scene/resources/skeleton_modification_2d_physicalbones.cpp:186:2: runtime error: member access within null pointer of type 'struct SkeletonModificationStack2D'
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on godotengine/godot/issues
[1] bin/godot.linuxbsd.tools.64s() [0x1e697d8] (/home/runner/work/godot/godot/platform/linuxbsd/crash_handler_linuxbsd.cpp:54)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7fd1ca5b0210] (??:0)
[3] SkeletonModification2DPhysicalBones::fetch_physical_bones() (/home/runner/work/godot/godot/scene/resources/skeleton_modification_2d_physicalbones.cpp:186)
[4] void call_with_variant_args_helper<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (/home/runner/work/godot/godot/./core/variant/binder_common.h:202 (discriminator 4))
[5] void call_with_variant_args_dv<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, int, Callable::CallError&, Vector<Variant> const&) (/home/runner/work/godot/godot/./core/variant/binder_common.h:315)
[6] MethodBindT<>::call(Object*, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/./core/object/method_bind.h:283)
[7] Object::call(StringName const&, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/core/object/object.cpp:784 (discriminator 1))
[8] Object::callv(StringName const&, Array const&) (/home/runner/work/godot/godot/core/object/object.cpp:709 (discriminator 1))
[9] void call_with_variant_args_ret_helper<__UnexistingClass, Variant, StringName const&, Array const&, 0ul, 1ul>(__UnexistingClass*, Variant (__UnexistingClass::*)(StringName const&, Array const&), Variant const**, Variant&, Callable::CallError&, IndexSequence<0ul, 1ul>) (/home/runner/work/godot/godot/./core/variant/binder_common.h:623 (discriminator 8))
[10] void call_with_variant_args_ret_dv<__UnexistingClass, Variant, StringName const&, Array const&>(__UnexistingClass*, Variant (__UnexistingClass::*)(StringName const&, Array const&), Variant const**, int, Variant&, Callable::CallError&, Vector<Variant> const&) (/home/runner/work/godot/godot/./core/variant/binder_common.h:399)
[11] MethodBindTR<Variant, StringName const&, Array const&>::call(Object*, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/./core/object/method_bind.h:452)
[12] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (/home/runner/work/godot/godot/modules/gdscript/gdscript_vm.cpp:1496 (discriminator 1))
[13] GDScriptInstance::call(StringName const&, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/modules/gdscript/gdscript.cpp:1552)
[14] Object::call(StringName const&, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/core/object/object.cpp:765 (discriminator 1))
[15] Variant::call(StringName const&, Variant const**, int, Variant&, Callable::CallError&) (/home/runner/work/godot/godot/core/variant/variant_call.cpp:938)
[16] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (/home/runner/work/godot/godot/modules/gdscript/gdscript_vm.cpp:1394)
[17] GDScriptInstance::call(StringName const&, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/modules/gdscript/gdscript.cpp:1552)
[18] ScriptInstance::call(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/runner/work/godot/godot/core/object/script_language.cpp:311)
[19] Node::_notification(int) (/home/runner/work/godot/godot/scene/main/node.cpp:147)
[20] Node::_notificationv(int, bool) (/home/runner/work/godot/godot/./scene/main/node.h:45 (discriminator 14))
[21] Object::notification(int, bool) (/home/runner/work/godot/godot/core/object/object.cpp:795)
[22] Node::_propagate_ready() (/home/runner/work/godot/godot/scene/main/node.cpp:189)
[23] Node::_set_tree(SceneTree*) (/home/runner/work/godot/godot/scene/main/node.cpp:2525)
[24] Node::_add_child_nocheck(Node*, StringName const&) (/home/runner/work/godot/godot/scene/main/node.cpp:1237)
[25] Node::add_child(Node*, bool) (/home/runner/work/godot/godot/scene/main/node.cpp:1250)
[26] void call_with_variant_args_helper<__UnexistingClass, Node*, bool, 0ul, 1ul>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool), Variant const**, Callable::CallError&, IndexSequence<0ul, 1ul>) (/home/runner/work/godot/godot/./core/variant/binder_common.h:202 (discriminator 4))
[27] void call_with_variant_args_dv<__UnexistingClass, Node*, bool>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool), Variant const**, int, Callable::CallError&, Vector<Variant> const&) (/home/runner/work/godot/godot/./core/variant/binder_common.h:315)
[28] MethodBindT<Node*, bool>::call(Object*, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/./core/object/method_bind.h:283)
[29] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (/home/runner/work/godot/godot/modules/gdscript/gdscript_vm.cpp:1496 (discriminator 1))
[30] GDScriptInstance::call(StringName const&, Variant const**, int, Callable::CallError&) (/home/runner/work/godot/godot/modules/gdscript/gdscript.cpp:1552)
[31] ScriptInstance::call(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/runner/work/godot/godot/core/object/script_language.cpp:311)
[32] Node::_notification(int) (/home/runner/work/godot/godot/scene/main/node.cpp:58)
[33] Node::_notificationv(int, bool) (/home/runner/work/godot/godot/./scene/main/node.h:45 (discriminator 14))
[34] CanvasItem::_notificationv(int, bool) (/home/runner/work/godot/godot/./scene/main/canvas_item.h:164 (discriminator 3))
[35] Control::_notificationv(int, bool) (/home/runner/work/godot/godot/./scene/gui/control.h:47 (discriminator 3))
[36] Object::notification(int, bool) (/home/runner/work/godot/godot/core/object/object.cpp:795)
[37] SceneTree::_notify_group_pause(StringName const&, int) (/home/runner/work/godot/godot/scene/main/scene_tree.cpp:809)
[38] SceneTree::process(float) (/home/runner/work/godot/godot/scene/main/scene_tree.cpp:443 (discriminator 2))
[39] Main::iteration() (/home/runner/work/godot/godot/main/main.cpp:2499)
[40] OS_LinuxBSD::run() (/home/runner/work/godot/godot/platform/linuxbsd/os_linuxbsd.cpp:261)
[41] bin/godot.linuxbsd.tools.64s(main+0x3ff) [0x1e67115] (/home/runner/work/godot/godot/platform/linuxbsd/godot_linuxbsd.cpp:60)
[42] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fd1ca5910b3] (??:0)
[43] bin/godot.linuxbsd.tools.64s(_start+0x2e) [0x1e66c5e] (??:?)
-- END OF BACKTRACE --
Aborted (core dumped)
FATAL ERROR: Godot has been crashed.
```

It might not be immediate to reproduce locally though, here's the config for the CI build that finds this crash: https://github.com/godotengine/godot/blob/master/.github/workflows/linux_builds.yml#L129-L170

Basically it builds Godot with `asan` and `ubsan`, and uses this build to edit and run https://github.com/qarmin/RegressionTestProject/archive/4.0.zip (the SwiftShader stuff is a software Vulkan renderer for CI, it shouldn't be needed to reproduce the issue).

**Comment by qarmin** [comment]
This last lines before crash:
```
SkeletonModification2DPhysicalBones.fetch_physical_bones
Parameters []
```
shows that probably this code is enough to reproduce crash
```
SkeletonModification2DPhysicalBones.new().fetch_physical_bones()
```

**Comment by TwistedTwigleg** [comment]
Thanks guys, I'll take a look at the issue once I have a chance :+1:

**Comment by fire** [comment]
Merging is blocked by:

```

#################### SkeletonModification2DPhysicalBones ####################
SkeletonModification2DPhysicalBones.set_physical_bone_chain_length
Parameters [100]
SkeletonModification2DPhysicalBones.get_physical_bone_chain_length
Parameters []
SkeletonModification2DPhysicalBones.set_physical_bone_node
Parameters [100, ]
ERROR: Joint index out of range!
   at: set_physical_bone_node (scene/resources/skeleton_modification_2d_physicalbones.cpp:259)
SkeletonModification2DPhysicalBones.get_physical_bone_node
Parameters [100]
ERROR: Joint index out of range!
   at: get_physical_bone_node (scene/resources/skeleton_modification_2d_physicalbones.cpp:265)
SkeletonModification2DPhysicalBones.fetch_physical_bones
Parameters []
scene/resources/skeleton_modification_2d_physicalbones.cpp:186:2: runtime error: member access within null pointer of type 'struct SkeletonModificationStack2D'
```

**Comment by TwistedTwigleg** [comment]
I fixed the build and documentation issues, as well as rebased with the latest version of master. 👍

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
